### DISCUSSION TITLE
How to train GraphSAGE with SparseTensor

### DISCUSSION THREAD

**Discussion by nt1tov** [pyg-team/pytorch_geometric#4317] [Q&A]
Is it possible to train graphSage with SparseTensor? I just changed type of reddit dataset to provide sparse Adj(SparseTensor class object for edges), but in first layer int forward(SageConv) after self.propagate I've got different size of data (out += self.lin_r(x_r). How to fix it?

**Comment by rusty1s** [comment]
Yes, this is definitely possible (usually by just swapping `edge_index` with `adj_t`. Can you show me an (minimal) example to reproduce?

**Reply by nt1tov** [reply]
I just took reddit.py example, replaced those lines from original:
```dataset = Reddit(path, transform=T.ToSparseTensor())```
 ```y_hat = model(batch.x, batch.adj_t.to(device))[:batch.batch_size] (in train function)```

**Reply by nt1tov** [reply]
but got the following:
```out += self.lin_r(x_r)```
```RuntimeError: The size of tensor a (232965) must match the size of tensor b (105702) at non-singleton dimension 0 ```

this line is from sage_conv.py

**Reply by nt1tov** [reply]
I'm getting different sizes for input and output of matmul method -- may be it is the problem

**Reply by rusty1s** [reply]
Thank you. I will check.

**Reply by nt1tov** [reply]
Please, share some feedback if you'll find some fix. My point is that sampled `adj_t `produces wrong size of output when it's used in `matmul `

**Reply by nt1tov** [reply]
By the way this problem also appears in backward -- I got the same problems:
` RuntimeError: Function torch::autograd::CppNode<SPMMMean> returned an invalid gradient at index 7 - got [232965, 256] but expected shape compatible with [105144, 256]`

where 105144 is the size of input and 232965 is the size of matmul(torch_sparse lib) output

**Reply by rusty1s** [reply]
Fixed in https://github.com/pyg-team/pytorch_geometric/pull/4320. Thanks for reporting!
